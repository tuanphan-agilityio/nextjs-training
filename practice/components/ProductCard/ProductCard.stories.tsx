import { Story, Meta } from '@storybook/react';

import ProductCard, { ProductCardProps } from '.';

export default {
  title: 'ProductCard',
  component: ProductCard,
} as Meta;

const Template: Story<ProductCardProps> = (args) => <ProductCard {...args} />;

export const Default = Template.bind({});
Default.args = {
  name: 'Sample Product',
  price: 100,
  status: 'In Stock',
  imgHref: '/images/product-1.jpg',
};
