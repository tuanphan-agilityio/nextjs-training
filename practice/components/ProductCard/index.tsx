import { FC, memo } from 'react';
import Image from 'next/image';
import Link from 'next/link';

import { ROUTES } from '@/constants/routes';

import RateIcon from '../../public/icons/rate.svg';

interface ProductCardProps {
  id: number;
  name: string;
  price: number;
  status: string;
  imgHref: string;
}

const ProductCard: FC<ProductCardProps> = ({
  id,
  name,
  price,
  status,
  imgHref,
}) => {
  return (
    <Link
      href={{
        pathname: ROUTES.PRODUCT,
        query: { productId: id },
      }}
    >
      <a className='block w-product-card sm:w-full'>
        <section className='bg-primary text-secondary shadow text-sm hover:shadow-xl hover:cursor-pointer'>
          <Image
            src={imgHref || '/images/default.png'}
            alt={name}
            width={'265px'}
            height={'248px'}
            priority
          />
          <div className='px-4 pt-4 pb-6 '>
            <div className='flex justify-between items-center mb-2 font-primary-bold xl:flex-wrap'>
              <h3>{name}</h3>
              <span>{price}$</span>
            </div>
            <p className='mb-3 text-tertiary text-start'>{status}</p>
            <div className='flex items-center gap-2'>
              <RateIcon />
              <span className='text-tertiary font-tertiary-regular'>(1k)</span>
            </div>
          </div>
        </section>
      </a>
    </Link>
  );
};

export type { ProductCardProps };

export default memo(ProductCard);
