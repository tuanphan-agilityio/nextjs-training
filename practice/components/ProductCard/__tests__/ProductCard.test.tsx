import { render } from '@testing-library/react';

import ProductCard, { ProductCardProps } from '..';

describe('ProductCard Component', () => {
  const mockProps: ProductCardProps = {
    id: 1,
    name: 'Product Name',
    price: 99.99,
    status: 'In Stock',
    imgHref: '/images/product-1.jpg',
  };

  it('renders correctly with mockProps', () => {
    const { container } = render(<ProductCard {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('displays the product name', () => {
    const { getByText } = render(<ProductCard {...mockProps} />);
    const productName = getByText('Product Name');
    expect(productName).toBeInTheDocument();
  });

  it('displays the product price', () => {
    const { getByText } = render(<ProductCard {...mockProps} />);
    const productPrice = getByText('99.99$');
    expect(productPrice).toBeInTheDocument();
  });

  it('displays the product status', () => {
    const { getByText } = render(<ProductCard {...mockProps} />);
    const productStatus = getByText('In Stock');
    expect(productStatus).toBeInTheDocument();
  });

  it('displays the product image', () => {
    const { getByAltText } = render(<ProductCard {...mockProps} />);
    const productImage = getByAltText('Product Name');
    expect(productImage).toBeInTheDocument();
  });
});
