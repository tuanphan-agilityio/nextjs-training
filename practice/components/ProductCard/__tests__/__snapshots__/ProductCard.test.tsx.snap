// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard Component renders correctly with mockProps 1`] = `
<div>
  <a
    class="block w-product-card sm:w-full"
    href="/1"
  >
    <section
      class="bg-primary text-secondary shadow text-sm hover:shadow-xl hover:cursor-pointer"
    >
      <div
        style="display: inline-block; max-width: 100%; overflow: hidden; position: relative; box-sizing: border-box; margin: 0px;"
      >
        <div
          style="box-sizing: border-box; display: block; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            role="presentation"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjY1IiBoZWlnaHQ9IjI0OCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
            style="max-width: 100%; display: block; margin: 0px; padding: 0px;"
          />
        </div>
        <noscript />
        <img
          alt="Product Name"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
      </div>
      <div
        class="px-4 pt-4 pb-6 "
      >
        <div
          class="flex justify-between items-center mb-2 font-primary-bold xl:flex-wrap"
        >
          <h3>
            Product Name
          </h3>
          <span>
            99.99
            $
          </span>
        </div>
        <p
          class="mb-3 text-tertiary text-start"
        >
          In Stock
        </p>
        <div
          class="flex items-center gap-2 flex-wrap"
        >
          <svg
            fill="none"
            height="18"
            width="106"
            xmlns="http://www.w3.org/2000/svg"
          >
            <g
              fill="#FFD43C"
            >
              <path
                d="M8.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561L1.697 7.127a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM52.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM30.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM74.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.5.5 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41ZM96.524 1.476a.5.5 0 0 1 .952 0l1.422 4.41a.5.5 0 0 0 .477.346l4.634-.01a.5.5 0 0 1 .294.905l-3.755 2.716a.501.501 0 0 0-.182.56l1.442 4.405a.5.5 0 0 1-.77.56l-3.743-2.733a.5.5 0 0 0-.59 0l-3.743 2.732a.5.5 0 0 1-.77-.56l1.442-4.403a.5.5 0 0 0-.182-.561l-3.755-2.716a.5.5 0 0 1 .294-.905l4.634.01a.5.5 0 0 0 .477-.346l1.422-4.41Z"
              />
            </g>
          </svg>
          <span
            class="text-tertiary font-tertiary-regular"
          >
            (1k)
          </span>
        </div>
      </div>
    </section>
  </a>
</div>
`;
